composer/utils/collect_env.py

```diff
@@ -0,0 +1,245 @@
+
+# Public function to restore original excepthook
+def restore_excepthook() -> None:
+    """Restore default :func:`sys.excepthook` behavior.
+
+    To restore the default :func:`sys.__excepthook__`:
+
+    .. testsetup::
+
+        import sys
+        from composer.utils.collect_env import restore_excepthook
+
+    .. doctest::
+        
+        >>> restore_excepthook()
+        >>> sys.excepthook
+        <built-in function excepthook>
+    """
+    sys.excepthook = sys.__excepthook__
+
+
+# Get Torch environment info
+def get_torch_env() -> str:
+    """Query Torch system environment via :mod:`torch.utils.collect_env`."""
+    return torchenv.get_pretty_env_info()
+
+
+# Composer environment information string output format
+composer_env_info_fmt = """
+Composer version: {composer_version}
+Number of nodes: {number_of_nodes}
+Host processor model name: {host_processor_model_name}
+Host processor core count: {host_processor_core_count}
+Accelerators per node: {accelerators_per_node}
+Accelerator model name: {accelerator_model_name}
+""".strip()
+
+
+# Get Composer environment info
+def get_composer_env() -> str:
+    """Query Composer pertinent system information."""
+
+    mutable_dict = ComposerEnv(
+        composer_version=get_composer_version(),
+        number_of_nodes=get_num_nodes(),
+        host_processor_model_name=get_host_processor_name(),
+        host_processor_core_count=get_host_processor_cores(),
+        accelerators_per_node=get_accel_per_node(),
+        accelerator_model_name=get_accel_model_name(),
+    )._asdict()
+
+    return composer_env_info_fmt.format(**mutable_dict)
+
+
+# Generate and print environment report
+def print_env() -> None:
+    """Generate system information report.
+
+    Example:
+
+    .. testsetup::
+
+        from composer.utils import print_env
+
+    .. doctest::
+
+        >>> print_env()
+        ---------------------------------
+        System Environment Report
+        ...
+    """
+
+    # Creation timestamp for report
+    creation_time = time.strftime('%Y-%m-%d %H:%M:%S %Z', time.localtime(time.time()))
+    report_hdr = f"---------------------------------\n" + \
+                 f"System Environment Report        \n" + \
+                 f"Created: {creation_time}\n"          + \
+                 f"---------------------------------\n"
+
+    # Torch section
+    print(report_hdr)
+    print("PyTorch information")
+    print("-------------------")
+    if TORCH_AVAILABLE:
+        # Only print Torch system info if installed
+        print(get_torch_env() + "\n\n")
+    else:
+        print("Torch not installed")
+
+    print("Composer information")
+    print("--------------------")
+    if COMPOSER_AVAILABLE:
+        # Only print Composer system info if installed
+        print(get_composer_env() + "\n\n")
+    else:
+        print("Composer not installed")
+
+
+# Invoke as standalone CLI script
+def main() -> None:
+    print("Collecting system information...")
+    print_env()
```

Comment:
(If someone is calling this as a standalone script, agree that it should go to stdout)